SI-6035: Specialization and separate compilation.

gkossakowski opened the pull request:
Fix an issue when specialization wouldn't work properly
in a setting where separate compilation was involved.
E.g. if you inherit from a class that is specialized and
has been compiled separately you might not get proper
forwarders generated. A test-case that this commit
adds covers that scenario.

The root cause of this issue was the fact that
SPECIALIZED flag was not pickled. Thanks to recent
Flags reorganization (see a9b85db) all that needs
to be done is to set the flag before pickling.

We choose to that in superaccessors phase because
it's a phase that runs before pickling and after
type checking (so we can check if given symbol
has an annotation).

Review by @axel22 and @odersky.

gkossakowski committed:
SI-6035: Specialization and separate compilation.

Fix an issue when specialization wouldn't work properly
in a setting where separate compilation was involved.
E.g. if you inherit from a class that is specialized and
has been compiled separately you might not get proper
forwarders generated. A test-case that this commit
adds covers that scenario.

The root cause of this issue was the fact that
SPECIALIZED flag was not pickled. Thanks to recent
Flags reorganization (see a9b85db) all that needs
to be done is to set the flag before pickling.

We choose to that in superaccessors phase because
it's a phase that runs before pickling and after
type checking (so we can check if given symbol
has an annotation).

Review by @axel22 and @odersky.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/515/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/515/

axel22 commented:
Btw:

https://github.com/gkossakowski/scala/blob/5e1e126a540ac2d5a7e067d396c1ae473b0f2cb6/src/compiler/scala/tools/nsc/transform/UnCurry.scala#L567

I think we can remove the check in `UnCurry` now, no?

VladUreche commented on src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala:
```diff
@@ -208,6 +212,11 @@ abstract class SuperAccessors extends transform.Transform with transform.TypingT
           mayNeedProtectedAccessor(sel, args, false)
 
+        // set a flag for all type parameters with `@specialized` annotation so it can be pickled
+        case typeDef: TypeDef if typeDef.symbol.hasAnnotation(definitions.SpecializedClass) =>
```
Greg, you set the SPECIALIZED flag here and it gets picked, but how do you know what value types the type param is specialized on? You need to read the annotation again, right? If so, why do you need the flag, couldn't you just read the annotation again?

gkossakowski replied:
Vlad, you are right. Both the flag and annotation gets pickled. Once you check the flag is set you read the annotation. The point of doing that is to have very quick check (by using a flag) for short-cutting. Does it make sense?

gkossakowski committed:
Remove `SPECIALIZED` flag setting in Uncurry.

The previous commit (5e1e126a) makes this code
not needed anymore because we set the flag in
Superaccessors.

gkossakowski commented:
@axel22: You are right. I totally forgot. The commit arrived.

axel22 commented:
lgtm!

gkossakowski commented:
Is jenkins going to recheck this pull request automatically?

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/522/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/522/

gkossakowski commented:
FYI: I'm looking into that failure now.

gkossakowski commented:
Adressed all feedback (including failing tests) in new pull request: https://github.com/scala/scala/pull/915

Tests were failing because I was setting a flag on a type parameter skolem and not type parameter itself.
